### ISSUE TITLE
imshow bug on GTKCairo and GTK3Cairo

### ISSUE THREAD

**Issue by OceanWolf** [matplotlib/matplotlib#4154]
On master, the following code fails (works fine on master with ``GTK3Agg`` and with ``GTK3Cairo`` on 1.3.1 (debian))
```python
import matplotlib
matplotlib.use('GTK3Cairo')
import matplotlib.pyplot as plt
import numpy as np

data = np.linspace(0, 10, 80).reshape(10, 8)

plt.imshow(data)
plt.show()
```

```
File "matplotlib/lib/matplotlib/backends/backend_cairo.py", line 169, in in draw_image
    buf, cairo.FORMAT_ARGB32, cols, rows, cols*4)
TypeError: Cannot use string as modifiable buffer
```

**Comment by tacaswell** [comment]
python 2 or 3?  My guess is that this is related to the str -> (byte, unicode) conversion + conversion using six.

**Comment by OceanWolf** [comment]
python2.7, using python-cairo version 1.8.8 on both mpl-1.3.1 and master.

**Comment by OceanWolf** [comment]
Ooh, debian has just upped, to mpl-1.4.2 on debian-testing...

Just tested on 1.4.2 (which uses six) and I don't see the problem, but the upgrade has also upgraded a few other things as well, and borked my access to my testing directory of python, so I shall try and fix that tomorrow and test again and come back with some news.

**Comment by OceanWolf** [comment]
Okay, tested it again and I still see it on master, fine using 1.4.2, bad on master, both using python 1.7.8 and python-cairo version 1.8.8.

**Comment by OceanWolf** [comment]
Okay, some poking in the code:

mpl_version | print(type(buf))
--------------- | -------------------
1.4.2            | ``<type 'buffer'>``
master         | ``<type 'str'>``

I will try and follow the code back further, to see why this happens.

**Comment by OceanWolf** [comment]
Okay, my skills run dead as this looks like a bug in the c to python code.  I have no idea about that layer of code.

**Comment by tacaswell** [comment]
This is probably related to the major cxx removal refactor then attn @mdboom

**Comment by mdboom** [comment]
Try #4163.

**Comment by OceanWolf** [comment]
I can't seem to pull this, I have weak git-foo...

``git fetch upstream pull/4154/head:pr4154``

**Comment by mdboom** [comment]
Use 4163.  4154 is just an issue (with no code attached).

**Comment by tacaswell** [comment]
Closed by 2a028c0
